### ISSUE TITLE
BeartypeDecorHintPepSignException and sub classing typing.Generic

### ISSUE THREAD

**Issue by eehusky** [beartype/beartype#29]
This is more of a Im not sure what I am looking at than a full up issue.  The code and exception I am having issues with follows.

```python
T = TypeVar('T', bound=Union['AtomicResults', Move])
U = TypeVar('U', bound='AtomicResults')

class FunctionDispatcher(Generic[T, U]):
    def __init__(self,
                 start: Callable[[T, T], U],
                 middle: Callable[[T, T, T], U],
                 stop: Callable[[T, T], U],
                 atomic: Callable[[T], U]) -> None:...

@beartype
def _run_step(
    incoming: Iterator[T], 
    func: FunctionDispatcher[T, U]
    ) -> Iterator[U]:...
# This function raises the following exception
beartype.roar.BeartypeDecorHintPepSignException: PEP 484 type hint newplanner.FunctionDispatcher[~T, ~U] representation "newplanner.FunctionDispatcher[~T, ~U]" not prefixed by "typing.".
```

Poking around the source code I found this section describing why it was rejected.
```python
    # If this representation is *NOT* prefixed by "typing,", this hint does
    # *NOT* originate from the "typing" module and is thus *NOT* PEP-compliant.
    # But by the validation above, this hint is PEP-compliant. Since this
    # invokes a world-shattering paradox, raise an exception
    if not sign_name.startswith('typing.'):
        raise BeartypeDecorHintPepSignException(
            f'PEP 484 type hint {repr(hint)} '
            f'representation "{sign_name}" not prefixed by "typing.".')
```

My reading of this tells me that beartype is looking for the type to come directly out of the typing module, which in my case the parent class is.  But the only way I know to instantiate a generic is to subclass it...so...does this imply that beartype does not support checking user defined subclasses of typing.Generic?  I would think that as long as the super class is from typing it would be okay.

Note that as long as i remove the @beartype annotation from the `_run_step` function all other interactions with this `FunctionDispatcher` class work fine.  I also tried various combinations of the definition to remove the forward references and bounds but to no avail and mypy doesnt seem to have any issues with it.  I did try typeguard out, it didnt have a problem with this function but I ran into some other problems so Im not sure if it was silently ignoring this one or actually handled it.

beartype 0.5.1
python 3.7.9

Thanks!
-Brandon

**Comment by leycec** [comment]
**Ugh!** Thanks a bundle of hibernating bears for the detailed writeup, Brandon. I humbly appreciate the inhuman effort in submitting detailed technical exegesis like this. It's *not* easy! You did great.

We, however, did not. `typing.Generic` is hell, so it's unsurprising but still disappointing that we fell down and couldn't get up when passed a parametrized generic. I'm currently disembowelling our generics handling with respect to subscripted generics under Python >= 3.9. <sup>*it is a special Hell*</sup>

Resolving this slots in nicely with that. In theory, I should have a fix by Monday or Tuesday and a new stable release shortly thereafter. Let's choose to believe what I am saying, because the alternative is unthinkable.

> beartype does not support checking user defined subclasses of typing.Generic

`beartype` *definitely* supports checking user-defined subclasses of `typing.Generic` – just not this one, for unknown reasons **that are absolutely our fault.** This isn't on you; this is on you. Generics are super-crazy to type-check (either at runtime or static time), because they come bundled with a bazillion different edge cases, outliers, and subtleties that make my eyes bleed crocodile tears.

I applaud mypy for properly supporting your hefty, plus-sized generic. :clap: 

Oddly, our test suite exercises a similar use case:

```
class Pep484GenericTypevaredSingle(Generic[S, T]):
    '''
    `PEP 484`_-compliant user-defined generic subclassing a single parametrized
    :mod:`typing` type.

    .. _PEP 484:
       https://www.python.org/dev/peps/pep-0484
    '''

    pass
```

That works, so yours should do it. But it doesn't, so we botched it bad. `</sigh>`

**Comment by leycec** [comment]
**Aha!** Your code is correct and valid, but we neglected to test an edge case associated with parametrized generics (i.e., generics subscripted by one or more type variables), because we currently just ignore parameters everywhere, except it seems that we just can't ignore parametrized generics.

If you'd like to **immediately** repair your code with respect to `beartype`, would you mind temporarily stripping the `[T, U]` subscription from `FunctionDispatcher`? That is, would you mind  temporarily reducing the above annotation to:

```
@beartype
def _run_step(
    incoming: Iterator[T], 
    func: FunctionDispatcher
    ) -> Iterator[U]:...
```

I can confirm that behaves as expected. Thus was sanity restored to the multiverse!

![Mind. Blown.](https://static1.srcdn.com/wordpress/wp-content/uploads/2020/11/tim-and-eric-mind-blown.jpg)

Subscripting `FunctionDispatcher` by `[T, U]` is *technically* redundant, because its superclass is already subscripted by the same parameters. So, there's no loss of type safety in dropping those parameters there.

Of course, it's obviously still bad that we raise non-human-readable exceptions on parametrized generics, because that's obviously a valid use case. I promise to resolve this shortly or my middle name isn't Bearhugger.

**Comment by eehusky** [comment]
Haha no worries, ever heard the saying that you can identify a test engineer by their overly detailed bug reports?  The exception was perfect and entirely readable!  What stuck in my craw about it was the 'prefiixed by typing.' comment and the implication that you couldn't inherent from generic classes which seemed nonsensical.

I feel your pain, I have been working on a toy project and decided this toy would be the time for me to learn about python type annotations, one home brewed re-implementation of dataclasses with hooks for runtime type checking later i started working on my own type checker.  Got about 20 minutes into it, recognized the pandoras box size can of worms I was walking into and promptly googled 'python runtime type checking library' and told Alexa to order some worm killer.

That being said, I have no doubt my plus sized type hints are a bit...well...they are beautiful on the inside okay bud?  But my understanding of how I thought it was working was roughly this and by all means correct me if my rational is out in the weeds.

```python
def plan_nd_blend_first(curr: AtomicResults, next: AtomicResults) -> BlendResults
def plan_nd_blend_middle(prev: AtomicResults, curr: AtomicResults, next: AtomicResults) -> BlendResults
def plan_nd_blend_last(prev: AtomicResults, curr: AtomicResults) -> BlendResults
def plan_nd_blend_atomic(curr: AtomicResults) -> BlendResults

blend_dispatch = FunctionDispatcher[AtomicResults, BlendResults](
    start=plan_nd_blend_first,
    middle=plan_nd_blend_middle,
    stop=plan_nd_blend_last,
    atomic=plan_nd_blend_atomic,
)

#error: Missing type parameters for generic type "FunctionDispatcher"
@beartype
def _run_step(
    incoming: Iterator[T], 
    func: FunctionDispatcher
    ) -> Iterator[U]:...

#error: Need type annotation for 'new_results'
new_results = _run_step(previous_stage_results, blend_dispatch ) 
# because the U parameter isnt getting bound in the definition of _run_step by the mapping of T->U in the function dispatcher class it cant infer that the return type of _run_step will be U given that it was passed an iterator of type T.

# If i add the [T,U] to the runstep function mypy is able to correctly infer that the type of new_result will be U
# Basically the [T,U] binds the Iterator[T] to the Iterator[U]? I think.
```

This is some relevant excerpts from the file in question and why 'I' think that you do need the type variables added to the type in the function parameters?  Or maybe more correctly, why mypy is telling me that I need to think that....so maybe my source of truth is off?  Who knows what to believe these days.

**Comment by leycec** [comment]
> What stuck in my craw about it was the `'prefixed by typing.'` comment...

*Yuppers.* That's a non-human-readable exception to be sure and something that wasn't really intended to be exposed to my delightful userbase. Of course, it was. I can only shake my head with shame and regret.

A bit of background on that exception, if you're curious. You're not, but you're getting the background anyway. In `beartype` speak, a "sign" is an arbitrary object used to uniquely identify the pseudo-type (i.e., category, kind, sort) of a type hint. I say "pseudo-type," because the `typing` module was designed in a balls cray-cray way that explicitly prohibits standard operations like `isinstance()` and `issubclass()` on type hints. That means we have to be even cleverer than our Maine Coon on a slippery snowday to distinguish type hints from one another. Here are a few examples of signs:

* `typing.List`, the sign of all list type hints (e.g., `[List[T]`, `List[List[List[complex]]]`).
* `typing.Union`, the sign of all union type hints (e.g., `Union[int, bool]`, `Union[List[str], Dict[float, U]]`).
* `typing.Generic`, the sign of all generic type hints (e.g., `FunctionDispatcher[T, U]`).

When we see a type hint in `beartype`, we first internally map that hint to its sign and then use that sign to generate highly optimized code checking arbitrary objects against that hint. The sign helps us "figure out" what we're dealing with. That usually works, because we have tests that exercise edge cases. When that doesn't work, because our tests sucked it up yet again, we self-flagellate ourselves with a burlap sack and raise a non-human-readable `BeartypeDecorHintPepSignException`. Basically, we give up, because we have failed at even the most basic type hint detection.

That's what you saw. We need to explicitly detect parametrized generics, but we weren't, so we had no idea that `FunctionDispatcher[T, U]` even *was* a generic. It is all awful!

> Got about 20 minutes into it, recognized the pandoras box size can of worms I was walking into and promptly googled 'python runtime type checking library' and told Alexa to order some worm killer.

*"And make it stat, Alexa! We've got worms the size of your android bobble head over here crawling all over my codebase and making creepy burbling noises from unmentionable body parts!"*

You do not want to walk into that box. Friends don't let friends type-check.

```
#error: Missing type parameters for generic type "FunctionDispatcher"
@beartype
def _run_step(
    incoming: Iterator[T], 
    func: FunctionDispatcher
    ) -> Iterator[U]:...
```

I'm *pretty* sure that's an unresolved mypy bug, just as I'm *pretty* sure they won't see it that way. `beartype` correctly infers the type parameters of `FunctionDispatcher` without compliant – because its superclass is already parametrized. In theory, static type checkers mypy *should* be able to do that too. In practice, mypy is... *particular*. Okay, it's opinionated. Okay, it's pedantic. Okay, it's a fundamentalist zealot currently on a three-year pilgrimage to the exalted shrine of Guido at Mountain Grove, California.

That's how particular mypy is. Mypy developers (and therefore mypy itself) have very fixed, very firm, and very unyielding opinions about how Python code should be type-checked.

Mypy isn't necessarily wrong here – but it isn't necessarily right, either. The *real* issue is that [PEP 484](https://www.python.org/dev/peps/pep-0484/#user-defined-generic-types) is unnecessarily vague on this topic and vagueness is the death of all good standards.

And... I should probably have just been fixing the issue rather than writing the diatribe. What you gonna do? It's a snow Saturday. A guy gotta write diatribes on snow Saturdays. :mountain_snow:

**Comment by leycec** [comment]
I've added a unit test exercising this edge case, which means tests are now failing. But that's good, because tests *should* have been failing for years but didn't. `</shaking_my_snow_covered_head>`

**Comment by eehusky** [comment]
That is an awesome explanation thank you!  I have to admit that your commitment to the bear theme is even more laudable than your excellent library! I just noticed the what I can only assume is intentional, type->sign~>bear sign~>possible bear berries reference and actually laughed out loud.

I completely understand about the what some might call...dogmatic...approach they take over there and agree some areas are extremely vague. (Ill gloss over how vague standards led to undefined behavior in C, it at least had the excuse of not knowing any better at the time, I doubt drawing parallels between the two would go over well.) 

I'm going to play devils advocate here for second if you'll play along.  Here are the internals of the run_step function we have been talking about.

```python
def _run_step(incoming: Iterator[T], func: FunctionDispatcher[T, U]) -> Iterator[U]:
    gen = _delayline(incoming, 3)
    for element in gen:
        is_start, is_middle, is_stop, is_atomic = _mark_ends(element)
        if is_atomic:
            _, curr, _ = element
            assert curr is not None
            yield func.atomic(curr)
        # more stuff....
```

Now I would agree that mypy should be able to figure this out if the function in question was used in isolation.

```python
def plan_trajectory(gen: Iterator[Move], n_axis: int) -> Iterator[BlendResults]:
        return _run_step(gen, atomic_dispatch) # at this point run step is passed Iterator[Move] and it knows that the return type shall be BlendResults.
```

But, it gets used in a data pipeline fashion roughly like this. (Don't get me started on mypys inability to pick up on re-using variable names.)  This is the context in which mypy cant pick out the return types without the type parameters.

```python
def plan_trajectory(gen: Iterator[Move], n_axis: int) -> Iterator[ExtractResults]:
        gen2 = _run_step(gen, atomic_dispatch) # at this point run step is passed Iterator[Move]
        gen3 = _run_step(gen2, blend_dispatch) # but without the parameters in the function dispatch thing it cant make the leap between here and the end.
        gen4 = _run_step(gen3, link_dispatch)
        gen5 = _run_step(gen4, extract_dispatch)
        gen6 = _run_step(gen5, check_nd_dispatch)
        return gen6 # mypy knows (can assume based on function signature) that gen6 should be Iterator[ExtractResults]
```

Now if I go and mix and match the order of the inner steps it wouldnt have any way to pick that out right?  Because without the static mapping defined by the dispatcher types gen3 could be a killer slug for all it knows?  Now since beartype doesnt match the parameters of generics (I think thats a correct statement) it shouldnt care that the parameters are missing, just that it is generic, so you would be correct in that regard.

*steps away from the computer and comes back*

Now you have got me wondering if you meant that if i redefined this class without the class level parameters.
```python
class FunctionDispatcher:
    def __init__(self,
                 start: Callable[[T, T], U],
                 stop: Callable[[T, T], U],
                 middle: Callable[[T, T, T], U],
                 atomic: Callable[[T], U]) -> None:...
```

That mypy should be able to work backwards from the declaration of these functions via just the parameters of the callables?
```python
def plan_nd_blend_first(curr: AtomicResults, next: AtomicResults) -> BlendResults: ...
def plan_nd_blend_middle(prev: AtomicResults, curr: AtomicResults, next: AtomicResults) -> BlendResults: ...
def plan_nd_blend_last(prev: AtomicResults, curr: AtomicResults) -> BlendResults: ...
def plan_nd_blend_atomic(curr: AtomicResults) -> BlendResults: ...
blend_dispatch = FunctionDispatcher[AtomicResults, BlendResults](
    start=plan_nd_blend_first,
    middle=plan_nd_blend_middle,
    stop=plan_nd_blend_last,
    atomic=plan_nd_blend_atomic,
)
```
When I wrote that class it just sort of looked right and seemed reasonable enough, I never actually tried completely removing it from the class definition.  Not sure if there might be a better "container" to group these things together than a class like that.  Im confident that if i passed the functions directly into the `_run_step` function mypy wouldnt have a problem with it.  All that dispatcher class does is keep me from having to keep track of the individual substeps....so Im not really attached to the implementation, just the result.

On the other hand the intent is that *all* the functions have the same type signature so that sort of enforces/documents that intent.  It makes sense/seems more reasonable if you come at it from that angle.

I'm currently digging into an unrelated bug in this project and when I exorcise that bugger ill comeback and try to appease the mypy demigods by sacrificing the generic class.

Enjoy the snow day!  I used to live the Michigan's UP and miss the Snow Saturdays!

**Comment by leycec** [comment]
Please find below this **photorealistic selfie** I captured shortly after resolving this issue in d15cba5.

![My head.](https://mediainreview.files.wordpress.com/2016/03/head-exploding.png?w=600&h=450)
<sup>*@leycec: he's looked better*</sup>

I have lived in a cabin for more years than a 4-byte register has representable decimal values. This is the first issue that reduced me to openly weeping on the cat fur-lined carpet. I'm not proud of that moment. I now have even less hair and smell slightly worse, which my wife didn't think was possible. It was.

I'm only proud that `beartype` now accepts your valid use case without complaint:

```
>>> from beartype import beartype
>>> from typing import TypeVar, Generic, Callable, Iterator, Union
>>> T = TypeVar('T', bound=Union['AtomicResults', bool])
>>> U = TypeVar('U', bound='AtomicResults')
>>> class FunctionDispatcher(Generic[T, U]):
...    def __init__(self,
...                 start: Callable[[T, T], U],
...                 middle: Callable[[T, T, T], U],
...                 stop: Callable[[T, T], U],
...                 atomic: Callable[[T], U]) -> None: pass
>>> @beartype
... def _run_step(
...     incoming: Iterator[T],
...     func: FunctionDispatcher,
... ) -> Iterator[U]:
...     print('Oh. My. Odin.')
...     return incoming
>>> iterator = iter({}.items())
>>> function_dispatcher = FunctionDispatcher(range, range, range, range)
>>> _run_step(iterator, function_dispatcher)
Oh. My. Odin.
```

**Odin approves.** :hammer: 

Anyone sane (so, us) would consider this a critical bug fix. I'll be releasing `beartype 0.6.0` shortly – hopefully by tomorrow evening, where "evening" is vaguely defined as "the sun has yet to rise somewhere on Earth." This is a bit sooner than expected, so `beartype 0.6.0` won't be quite as feature-full as we'd all hoped. Nonetheless, it will work. It will also have [full PEP 561 compliance](https://www.python.org/dev/peps/pep-0561), which will be nice for you. Working >>>> non-working.

En route to resolving this issue, I fully intended to do deadly conversational battle with your devil's advocate vis-a-vis a deep-dive on back propagation as it relates to the modern state of subscripted PEP-compliant generics. Then I realized no one would know or care what I was gibbering incoherently about in the corner and that I should probably just resolve this issue instead, which in hindsight I only partially regret. Oh, the spirited intellectual debates we might have had – *all gone, like arguments in the muddy snowdrifts of the lower U.P.*

Thanks again for all the vigorous discussion, detailed report, and continual for**bear**ance <sup>*...get it?*</sup> with unseemly `beartype` breakage. I hope you haven't dropped `beartype` support just yet – but I'd fully understand if you had.

You rock, Husky Michigan (Wo)Man! :dog:

**Comment by eehusky** [comment]
Haha, been there before but its fantastic you sorted it out!  Just pulled the new version and it appears to be happy again.  Wouldn't dream of dropping it and thanks again!
